src/cmap/connection_pool.ts

```diff
@@ -642,44 +653,49 @@ function processWaitQueue(pool: ConnectionPool) {
     }
   }
 
-  const maxPoolSize = pool.options.maxPoolSize;
-  if (pool.waitQueueSize && (maxPoolSize <= 0 || pool.totalConnectionCount < maxPoolSize)) {
+  const { maxPoolSize, maxConnecting } = pool.options;
+  while (
+    pool.waitQueueSize &&
+    pool.pendingConnectionCount < maxConnecting &&
+    pool.availableConnectionCount <= 0 &&
+    (maxPoolSize <= 0 || pool.totalConnectionCount < maxPoolSize)
+  ) {
+    const waitQueueMember = pool[kWaitQueue].shift();
+    if (!waitQueueMember || waitQueueMember[kCancelled]) {
+      continue;
+    }
     createConnection(pool, (err, connection) => {
-      const waitQueueMember = pool[kWaitQueue].shift();
-      if (!waitQueueMember || waitQueueMember[kCancelled]) {
+      if (waitQueueMember[kCancelled]) {
         if (!err && connection) {
           pool[kConnections].push(connection);
           pool[kPending]--;
         }
 
-        pool[kProcessingWaitQueue] = false;
-        return;
-      }
-
-      if (err) {
-        pool.emit(
-          ConnectionPool.CONNECTION_CHECK_OUT_FAILED,
-          new ConnectionCheckOutFailedEvent(pool, err)
-        );
-      } else if (connection) {
-        pool[kCheckedOut]++;
-        pool[kPending]--;
-        pool.emit(
-          ConnectionPool.CONNECTION_CHECKED_OUT,
-          new ConnectionCheckedOutEvent(pool, connection)
-        );
-      }
+        // return;
+      } else {
+        if (err) {
+          pool.emit(
+            ConnectionPool.CONNECTION_CHECK_OUT_FAILED,
+            new ConnectionCheckOutFailedEvent(pool, err)
+          );
+        } else if (connection) {
+          pool[kCheckedOut]++;
+          pool[kPending]--;
+          pool.emit(
+            ConnectionPool.CONNECTION_CHECKED_OUT,
+            new ConnectionCheckedOutEvent(pool, connection)
+          );
+        }
```

Comment:
This is one of those code structures caused by TS. err and connection are mutually exclusive, we should have written something more like:
```js
if (err || !connection) {
  // err logic
  return
}
```
maybe that's why the early return looked erroneous?

Reply:
no, the early return (originally L656) was erroneous because if we encounter a timed out queue member we wouldn't try to process anymore of the waitqueue; ostensibly, if a queue member timed out, any remaining queue members are also timed out, but that doesn't mean we should just leave them sitting in the queue until some other event re-triggers its processing.